### ISSUE TITLE
tracepoint args with multiple attach points

### ISSUE THREAD

**Issue by brendangregg** [iovisor/bpftrace#132]
Perhaps one, or both, of these should work:

```
# bpftrace -e 'tracepoint:syscalls:sys_exit_statfs,tracepoint:syscalls:sys_exit_statx { @[args->ret] = count(); }'
The args builtin only works for probes with 1 attach point

# bpftrace -e 'tracepoint:syscalls:sys_exit_stat* { @[args->ret] = count(); }'
No such file or directory: /sys/kernel/debug/tracing/events/syscalls/sys_exit_stat*/format
Struct/union of type '_tracepoint_syscalls_sys_exit_stat*' does not contain a field named 'ret'
```

(Ignore the "No such file or directory" warning on wildcards; I'll file a separate bug for that: #133 )

There's multiple workarounds: you can list the probes one by one with the same code block for each. Or, you can use `ctx` and its offset from the /sys/kernel/debug/tracing/events/.../format file, eg:

```
# bpftrace -e 'tracepoint:syscalls:sys_exit_stat* { @[*(ctx + 16)] = count(); }'
No such file or directory: /sys/kernel/debug/tracing/events/syscalls/sys_exit_stat*/format
Attaching 2 probes...
^C

@[-2]: 2
```

So... should `args` work with mulitple attach points (probes)? `args->whatever` could mean different things logically, and matching them all with a wildcard may not make sens. They could also originate from different offsets.

I'm not adding a bug tag because I'm not sure yet if this is a bug, or is working as intended. Although I suspect people will try matching multiple tracepoints and will expect `args->ret` to be the return value.

**Comment by brendangregg** [comment]
changing to 0.9. I've seen newcomers run into this immediately, since they begin with:

```
bpftrace -e 'tracepoint:syscalls:sys_enter_open { printf("%s %s\n", comm, str(args->filename)); }'
```

(Maybe since it's the first one-liner in README.md), then discover it doesn't trace most opens, so then try:

```
bpftrace -e 'tracepoint:syscalls:sys_enter_open,tracepoint:syscalls:sys_enter_openat { printf("%s %s\n", comm, str(args->filename)); }'
```

and run into this issue.

**Comment by brendangregg** [comment]
@boat0 has written some code that begins to fix this issue (from #367):

```diff
--- tracepoint_format_parser.cpp	2019-01-17 11:34:12.503622806 -0800
+++ tracepoint_format_parser.cpp.1	2019-01-17 11:34:42.121113567 -0800
@@ -1,6 +1,7 @@
 #include <fstream>
 #include <iostream>
 #include <string.h>
+#include <glob.h>
 
 #include "ast.h"
 #include "struct.h"
@@ -30,23 +31,36 @@
         std::string &category = ap->target;
         std::string &event_name = ap->func;
         std::string format_file_path = "/sys/kernel/debug/tracing/events/" + category + "/" + event_name + "/format";
-        std::ifstream format_file(format_file_path.c_str());
 
-        if (format_file.fail())
-        {
-          std::cerr << "ERROR: tracepoint not found: " << ap->target << ":" << ap->func << std::endl;
-          // helper message:
-          if (ap->target == "syscall")
-            std::cerr << "Did you mean syscalls:" << ap->func << "?" << std::endl;
-          if (bt_verbose) {
-            if (format_file_path.find('*') == std::string::npos)
+        glob_t glob_result;
+        memset(&glob_result, 0, sizeof(glob_result));
+        int ret = glob(format_file_path.c_str(), 0, NULL, &glob_result);
+        if (ret != 0) {
+          if (ret == GLOB_NOMATCH) {
+            std::cerr << "ERROR: tracepoint not found: " << ap->target << ":" << ap->func << std::endl;
+            // helper message:
+            if (ap->target == "syscall")
+              std::cerr << "Did you mean syscalls:" << ap->func << "?" << std::endl;
+            if (bt_verbose)
               std::cerr << strerror(errno) << ": " << format_file_path << std::endl;
+          } else {
+            std::cerr << strerror(errno) << std::endl;
           }
 
+          globfree(&glob_result);
           return false;
         }
 
-        program->c_definitions += get_tracepoint_struct(format_file, category, event_name);
+        for (int i = 0; i < glob_result.gl_pathc; ++i) {
+          std::string filename(glob_result.gl_pathv[i]);
+          std::ifstream format_file(filename);
+          std::string prefix("/sys/kernel/debug/tracing/events/" + category + "/");
+          std::string real_event = filename.substr(prefix.length(),
+                  filename.length() - std::string("/format").length() - prefix.length());
+
+          program->c_definitions += get_tracepoint_struct(format_file, category, real_event);
+        }
+        globfree(&glob_result);
       }
     }
   }
```

That lets tracepoint_format_parser.cpp process wildcards. Further changes needed include only doing this if `args` is used.

**Comment by brendangregg** [comment]
see #370

**Comment by brendangregg** [comment]
fixed by #370
